csrc/sparse/cutlass/sparse_compressor.cu

```diff
@@ -0,0 +1,181 @@
+#include <cudaTypedefs.h>
+
+#include <torch/all.h>
+
+#include <ATen/cuda/CUDAContext.h>
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include "cutlass/cutlass.h"
+
+#include "cute/tensor.hpp"
+#include "cute/atom/mma_atom.hpp"
+#include "cutlass/numeric_types.h"
+#include "cutlass/numeric_conversion.h"
+#include "cutlass/detail/dependent_false.hpp"
+
+#include "cutlass_extensions/epilogue/broadcast_load_epilogue_c3x.hpp"
+#include "common.hpp"
+
+#include "cutlass/transform/device/transform_universal_adapter.hpp"
+#include "cutlass/transform/kernel/sparse_gemm_compressor.hpp"
+
+#include "cutlass/epilogue/collective/default_epilogue.hpp"
+#include "cutlass/epilogue/thread/linear_combination.h"
+#include "cutlass/gemm/collective/collective_builder.hpp"
+#include "cutlass/gemm/device/gemm_universal_adapter.h"
+#include "cutlass/gemm/kernel/gemm_universal.hpp"
+
+#include <iostream>
+
+#include "cutlass/cutlass.h"
+
+#include "cutlass/tensor_ref.h"
+#include "cutlass/epilogue/collective/collective_builder.hpp"
+#include "cutlass/gemm/dispatch_policy.hpp"
+
+#include "cutlass/util/host_tensor.h"
+#include "cutlass/util/packed_stride.hpp"
+
+#include "sparse_scaled_mm_c3x.cuh"
+
+/// Make A structured sparse by replacing elements with 0 and compress it
+template <typename ElementA_>
+bool sparsify_and_compress(torch::Tensor& a_compressed, torch::Tensor& e,
+                           torch::Tensor const& a) {
+  // Checks for conformality
+  TORCH_CHECK(a.dtype() == torch::kInt8 || a.dtype() == torch::kFloat8_e4m3fn ||
+              a.dtype() == torch::kFloat16 || a.dtype() == torch::kBFloat16);
+  TORCH_CHECK(a.dim() == 2)
+  // Check for strides and alignment
+  TORCH_CHECK(a.stride(1) == 1)
+
+  int m = a.size(0);
+  int k = a.size(1);
+
+  using ProblemShape = Shape<int, int, int, int>;
+  using ElementA = ElementA_;
+  using LayoutTagA = cutlass::layout::RowMajor;
+
+  // Layouts for reference (non-sparse) tensors
+  using StrideA = cutlass::gemm::TagToStrideA_t<LayoutTagA>;
+  using StrideE = StrideA;
+
+  using Gemm = typename std::conditional<
```

Comment:
I think we could replace this chain of conditionals with template specialization, i.e.

```
diff --git a/csrc/sparse/cutlass/sparse_compressor.cu b/csrc/sparse/cutlass/sparse_compressor.cu
index eac71bada..c9513a406 100644
--- a/csrc/sparse/cutlass/sparse_compressor.cu
+++ b/csrc/sparse/cutlass/sparse_compressor.cu
@@ -63,23 +63,8 @@ bool sparsify_and_compress(torch::Tensor& a_compressed, torch::Tensor& e,
   using StrideA = cutlass::gemm::TagToStrideA_t<LayoutTagA>;
   using StrideE = StrideA;
 
-  using Gemm = typename std::conditional<
-      std::is_same_v<ElementA, int8_t>,
-      typename sm90_int8_config_default<int8_t, cutlass::half_t,
-                                        ScaledEpilogue>::Cutlass3xGemm,
-      typename std::conditional<
-          std::is_same_v<ElementA, cutlass::float_e4m3_t>,
-          typename sm90_fp8_config_default<cutlass::float_e4m3_t,
-                                           cutlass::half_t,
-                                           ScaledEpilogue>::Cutlass3xGemm,
-          typename std::conditional<
-              std::is_same_v<ElementA, cutlass::half_t>,
-              typename sm90_fp16_config_default<cutlass::half_t,
-                                                cutlass::half_t,
-                                                ScaledEpilogue>::Cutlass3xGemm,
-              typename sm90_bf16_config_default<
-                  cutlass::bfloat16_t, cutlass::half_t,
-                  ScaledEpilogue>::Cutlass3xGemm>::type>::type>::type;
+  using Gemm = typename sm90_config_default<ElementA, cutlass::half_t,
+                                            ScaledEpilogue>::Cutlass3xGemm;
 
   using ElementAB = typename Gemm::ElementAB;
   using ElementD = typename Gemm::ElementD;
diff --git a/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu b/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu
index 1c67e2549..1ab3755d2 100644
--- a/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu
+++ b/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu
@@ -42,8 +42,7 @@ void cutlass_gemm_sm90_fp8_dispatch(torch::Tensor& out, torch::Tensor const& a,
   TORCH_CHECK(b.dtype() == torch::kFloat8_e4m3fn);
 
   using Cutlass3xGemmDefault =
-      typename sm90_fp8_config_default<InType, OutType,
-                                       Epilogue>::Cutlass3xGemm;
+      typename sm90_config_default<InType, OutType, Epilogue>::Cutlass3xGemm;
   using Cutlass3xGemmM64 =
       typename sm90_fp8_config_M64<InType, OutType, Epilogue>::Cutlass3xGemm;
   using Cutlass3xGemmM128 =
@@ -148,8 +147,7 @@ void cutlass_gemm_sm90_fp16_dispatch(torch::Tensor& out, torch::Tensor const& a,
   TORCH_CHECK(b.dtype() == torch::kFloat16);
 
   using Cutlass3xGemmDefault =
-      typename sm90_fp16_config_default<InType, OutType,
-                                        Epilogue>::Cutlass3xGemm;
+      typename sm90_config_default<InType, OutType, Epilogue>::Cutlass3xGemm;
 
   // m in (128, inf)
   return cutlass_sparse_gemm_caller<Cutlass3xGemmDefault>(
@@ -169,8 +167,7 @@ void cutlass_gemm_sm90_bf16_dispatch(torch::Tensor& out, torch::Tensor const& a,
   TORCH_CHECK(b.dtype() == torch::kBFloat16);
 
   using Cutlass3xGemmDefault =
-      typename sm90_bf16_config_default<InType, OutType,
-                                        Epilogue>::Cutlass3xGemm;
+      typename sm90_config_default<InType, OutType, Epilogue>::Cutlass3xGemm;
 
   // m in (128, inf)
   return cutlass_sparse_gemm_caller<Cutlass3xGemmDefault>(
@@ -190,8 +187,7 @@ void cutlass_gemm_sm90_int8_dispatch(torch::Tensor& out, torch::Tensor const& a,
   TORCH_CHECK(b.dtype() == torch::kInt8);
 
   using Cutlass3xGemmDefault =
-      typename sm90_int8_config_default<InType, OutType,
-                                        Epilogue>::Cutlass3xGemm;
+      typename sm90_config_default<InType, OutType, Epilogue>::Cutlass3xGemm;
   using Cutlass3xGemmM128 =
       typename sm90_int8_config_M128<InType, OutType, Epilogue>::Cutlass3xGemm;
   using Cutlass3xGemmM64 =
diff --git a/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cuh b/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cuh
index a948a02e4..084c99437 100644
--- a/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cuh
+++ b/csrc/sparse/cutlass/sparse_scaled_mm_c3x.cuh
@@ -481,29 +481,31 @@ void cutlass_sparse_gemm_caller(torch::Tensor& out, torch::Tensor const& a,
 
 template <typename InType, typename OutType,
           template <typename, typename, typename> typename Epilogue>
-struct sm90_fp16_config_default {
+struct sm90_config_default {};
+
+template <typename OutType,
+          template <typename, typename, typename> typename Epilogue>
+struct sm90_config_default<half_t, OutType, Epilogue> {
   // M in (128, inf)
-  static_assert(std::is_same<InType, cutlass::half_t>());
   using KernelSchedule = cutlass::gemm::KernelTmaWarpSpecializedPingpong;
   using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
   using TileShape = Shape<_128, _128, _128>;
   using ClusterShape = Shape<_2, _1, _1>;
   using Cutlass3xGemm =
-      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
+      cutlass_3x_gemm<half_t, OutType, Epilogue, TileShape, ClusterShape,
                       KernelSchedule, EpilogueSchedule, float>;
 };
 
-template <typename InType, typename OutType,
+template <typename OutType,
           template <typename, typename, typename> typename Epilogue>
-struct sm90_bf16_config_default {
+struct sm90_config_default<bfloat16_t, OutType, Epilogue> {
   // M in (128, inf)
-  static_assert(std::is_same<InType, cutlass::bfloat16_t>());
   using KernelSchedule = cutlass::gemm::KernelTmaWarpSpecializedPingpong;
   using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
   using TileShape = Shape<_128, _128, _128>;
   using ClusterShape = Shape<_2, _1, _1>;
   using Cutlass3xGemm =
-      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
+      cutlass_3x_gemm<bfloat16_t, OutType, Epilogue, TileShape, ClusterShape,
                       KernelSchedule, EpilogueSchedule, float>;
 };
 
@@ -621,18 +623,17 @@ struct sm90_fp8_config_8 {
 };
 ////////////////////////////////////////////////////////////////////////
 
-template <typename InType, typename OutType,
+template <typename OutType,
           template <typename, typename, typename> typename Epilogue>
-struct sm90_fp8_config_default {
+struct sm90_config_default<cutlass::float_e4m3_t, OutType, Epilogue> {
   // M in (128, inf)
-  static_assert(std::is_same<InType, cutlass::float_e4m3_t>());
   using KernelSchedule = cutlass::gemm::KernelTmaWarpSpecializedFP8FastAccum;
   using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
   using TileShape = Shape<_128, _128, _128>;
   using ClusterShape = Shape<_1, _2, _1>;
   using Cutlass3xGemm =
-      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
-                      KernelSchedule, EpilogueSchedule, float>;
+      cutlass_3x_gemm<cutlass::float_e4m3_t, OutType, Epilogue, TileShape,
+                      ClusterShape, KernelSchedule, EpilogueSchedule, float>;
 };
 
 template <typename InType, typename OutType,
@@ -709,19 +710,18 @@ struct sm90_fp8_config_M512 {
                       KernelSchedule, EpilogueSchedule, float, TileSchedule>;
 };
 
-template <typename InType, typename OutType,
+template <typename OutType,
           template <typename, typename, typename> typename Epilogue>
-struct sm90_int8_config_default {
+struct sm90_config_default<cutlass::int8_t, OutType, Epilogue> {
   // For M > 128 and any N
-  static_assert(std::is_same<InType, int8_t>());
   using KernelSchedule =
       typename cutlass::gemm::KernelTmaWarpSpecializedPingpong;
   using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
   using TileShape = Shape<_128, _128, _128>;
   using ClusterShape = Shape<_2, _1, _1>;
   using Cutlass3xGemm =
-      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
-                      KernelSchedule, EpilogueSchedule, int32_t>;
+      cutlass_3x_gemm<cutlass::int8_t, OutType, Epilogue, TileShape,
+                      ClusterShape, KernelSchedule, EpilogueSchedule, int32_t>;
 };
 
 template <typename InType, typename OutType,
```